### PR TITLE
TST: add timeouts for github actions tests and wheel builds.

### PR DESCRIPTION
**Pull request by ngoldbaum**
This is an attempt to track down the hang reported in #27872.

`pytest-timeout` sets a per-test timeout that also prints a C stack trace if it kills a timed out test runner. It looks like a few years ago there were issues using pytest-timeout and pytest-xdist together, but they seem to have been fixed.

I added it to `test-requirements.txt` since it's a pure-python dependency and should be harmless even if it's not needed. I can also just `pip install` it in the wheel testing script if that would be preferred.

I'm passing `--durations` and `--timeout` for all the wheel builds because it seems useful to me to always have this information and to always timeout if a single test takes longer than 30 minutes. I can also just do it for the free-threaded build if reviewers would prefer only doing it there.

### CODE REVIEW

File: numpy/linalg/tests/test_linalg.py [hunk 1]
```diff
@@ -1944,9 +1944,13 @@ def test_generalized_raise_multiloop():
 
     assert_raises(np.linalg.LinAlgError, np.linalg.inv, x)
 
+
 @pytest.mark.skipif(
     threading.active_count() > 1,
     reason="skipping test that uses fork because there are multiple threads")
+@pytest.mark.skipif(
+    NOGIL_BUILD,
+    reason="Cannot safely use fork in tests on the free-threaded build")
```
Comment (ngoldbaum) [inline numpy/linalg/tests/test_linalg.py:1953]: I noticed [in this CI run](https://github.com/numpy/numpy/actions/runs/12128567744/job/33815225281#step:8:1357) that despite the skip above, CPython was still warning about an unsafe use of fork, so I think it's probably safer to just always skip this test on the free-threaded build.

**Comment by charris** [comment]
I reran the hanging test today, and it passed. The problem may have been upstream.

**Comment by ngoldbaum** [comment]
IMO it's worth having these timeouts for heisenbugs and other infrequent hangs. Also `--durations` to see slow tests if they do complete.

**Review by rgommers — APPROVED** [review]
Looks like a good idea, +1 in principle. Note that the first job I checked was the `cp313t` smoke test in `linux.yml`, and `pytest-timeout` didn't actually show up. If you want it there, please add it in `.github/meson_actions/action.yml`. If not, please go ahead and merge I'd say.

**Comment by ngoldbaum** [comment]
Added a 10 minute timeout and `--durations=10` to most of the pytest invocations that happen on Github actions. I still think a 30 minute timeout makes sense for the wheel builds because pypy in particular can be very slow.

**Comment by ngoldbaum** [comment]
Awesome, CI is passing. I didn't re-trigger all the wheel builds because I haven't touched them since they passed in [9169099](https://github.com/numpy/numpy/pull/27899/commits/916909955202ad30b569aa59fc9fff46ed3cdf5c).

**Comment by charris** [comment]
> pypy in particular can be very slow.

Takes close to 90 minutes on windows. Some of the other builds are variable, sometimes verging on 50 minutes.

**Comment by ngoldbaum** [comment]
Let's pull this in. Please ping me if anyone notices spurious timeouts or if the timeouts prove to be unhelpful.

**Comment by charris** [comment]
Thanks Nathan, lets give this a shot. We will need to track the nightlies to see if any problems come up.

**Pull request merged**